Blizzard.IO.RabbitMQ/NetqRabbitConsumer.cs

```diff
@@ -0,0 +1,100 @@
+using Blizzard.IO.Core;
+using Blizzard.IO.RabbitMQ.Entities;
+using EasyNetQ;
+using EasyNetQ.Topology;
+using Microsoft.Extensions.Logging;
+
+namespace Blizzard.IO.RabbitMQ
+{
+    public class NetqRabbitConsumer<TData> : IConsumer<TData>, IConsumerWithMetadata<TData, RabbitMessageProperties>
+    {
+        private readonly IBus _netqBus;
+        private readonly IDeserializer<TData> _deserializer;
+        private readonly IAbstractTypeDeserializer<TData> _abstractDeserializer;
+        private readonly IQueue _netqQueue;
+        private readonly IConverter<MessageProperties, RabbitMessageProperties> _messagePropertiesToRabbitMessagePropertiesConverter;
+        private readonly ILogger _logger;
+        
+        private IDisposable _consumeHandler;
+        private readonly Func<byte[], MessageProperties, TData> _dataExtractionStrategy;
+
+        public event Action<TData> MessageReceived;
+        public event Action<TData, RabbitMessageProperties> MessageWithMetadataReceived;
+
+        protected NetqRabbitConsumer(IBus netqBus, RabbitQueue sourceQueue, ILoggerFactory loggerFactory,
+            IConverter<MessageProperties, RabbitMessageProperties> messagePropertiesToRabbitMessagePropertiesConverter)
+        {
+            _netqBus = netqBus;
+            _netqQueue = new Queue(sourceQueue.Name, sourceQueue.Exclusive);
+            _logger = loggerFactory.CreateLogger(nameof(NetqRabbitConsumer<TData>));
+            _messagePropertiesToRabbitMessagePropertiesConverter = messagePropertiesToRabbitMessagePropertiesConverter ?? new RabbitPropertiesConverter();
+        }
+
+        public NetqRabbitConsumer(IBus netqBus, RabbitQueue sourceQueue, IDeserializer<TData> deserializer, ILoggerFactory loggerFactory,
+            IConverter<MessageProperties, RabbitMessageProperties> messagePropertiesToRabbitMessagePropertiesConverter = null)
+            : this(netqBus, sourceQueue, loggerFactory, messagePropertiesToRabbitMessagePropertiesConverter)
+        {
+            _deserializer = deserializer;
+            _dataExtractionStrategy += ExtractConcreteData;
+            _logger.LogDebug($"Initialized consumer from queue: {_netqQueue.Name}," +
+                $" expecting {typeof(TData)} data");
+        }
+
+        public NetqRabbitConsumer(IBus netqBus, RabbitQueue sourceQueue, IAbstractTypeDeserializer<TData> abstractDeserializer,
+            ILoggerFactory loggerFactory,
+            IConverter<MessageProperties, RabbitMessageProperties> messagePropertiesToRabbitMessagePropertiesConverter = null)
+            : this(netqBus, sourceQueue, loggerFactory, messagePropertiesToRabbitMessagePropertiesConverter)
+        {
+            _abstractDeserializer = abstractDeserializer;
+            _dataExtractionStrategy += ExtractAbstractData;
+            _logger.LogDebug($"Initialized consumer from queue: {_netqQueue.Name}," +
+                $" expecting {typeof(TData)} derieved data");
+        }
+
+        public void Start()
```

Comment:
Add a way to prevent the user from calling start twice (we already talked about it)
throw exception if he is calling another time